### ISSUE TITLE
Uploading large body logs exception in stderr, mangles input

### ISSUE THREAD

**Issue by kubukoz** [dream11/zio-http#36]
Using the code:

```scala
import zhttp.http._
import zhttp.service.Server
import zio._

object Main extends App {
  val echo = Http.collect[Request] { case req @ Method.POST -> Root / "echo" =>
    Response.http(content = req.data.content)
  }

  override def run(args: List[String]): URIO[ZEnv, ExitCode] =
    Server.start(8090, echo).exitCode
}
```

Edit: made some adjustments after finding out #35 was an incorrect report.

```bash
# 10 million bytes
> cat /dev/random | head -c 10000000 > in.zip

> curl localhost:8090/echo --data-binary @in.zip -v > out.zip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8090 (#0)
> POST /echo HTTP/1.1
> Host: localhost:8090
> User-Agent: curl/7.64.1
> Accept: */*
> Content-Length: 10000000
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
} [65536 bytes data]
* We are completely uploaded and fine
< HTTP/1.1 200 OK
< content-length: 9469184
<
{ [102400 bytes data]
100 18.5M  100 9247k  100 9765k  32.7M  34.5M --:--:-- --:--:-- --:--:-- 67.2M
* Connection #0 to host localhost left intact
* Closing connection 0

# output is smaller
> wc -c in.zip                                                                                                                                                                                                            
 10000000 in.zip
> wc -c out.zip
 9469184 out.zip

shasum --algorithm 256 *.zip
8e27ef091cef676413a1e39f002a1fe5fafaa7ac4f80a2bb41a3b8092837bf52  in.zip
86807e45c1a2c61d4545954e62d6da7b1504691454ea593cfee79af61eba33b1  out.zip
```

```
[E] Mar 16, 2021 2:22:47 AM io.netty.channel.DefaultChannelPipeline onUnhandledInboundException
[E] WARNING: An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
[E] io.netty.channel.StacklessClosedChannelException
[E] 	at io.netty.channel.AbstractChannel.close(ChannelPromise)(Unknown Source)
[E]
```

Sending a 960MB file, however, hangs around the end (client-side) after outputting this from curl:

```
 % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8090 (#0)
> POST /echo HTTP/1.1
> Host: localhost:8090
> User-Agent: curl/7.64.1
> Accept: */*
> Content-Length: 1000000000
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
} [65536 bytes data]
 99  953M    0     0   99  950M      0  25.5M  0:00:37  0:00:37 --:--:-- 13.5M* We are completely uploaded and fine
100  953M    0     0  100  953M      0  21.1M  0:00:45  0:00:45 --:--:--     0^C #I cancelled here
```

<details>
<summary>After a couple seconds from disconnecting I got this stack trace</summary>


```
[E] Mar 16, 2021 2:55:54 AM io.netty.channel.DefaultChannelPipeline onUnhandledInboundException
[E] WARNING: An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
[E] java.lang.IndexOutOfBoundsException: writerIndex(0) + minWritableBytes(-1454240449) exceeds maxCapacity(2147483647): UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf(ridx: 0, widx: 0, cap: 1810942847)
[E] 	at io.netty.buffer.AbstractByteBuf.ensureWritable0(AbstractByteBuf.java:296)
[E] 	at io.netty.buffer.ByteBufUtil.reserveAndWriteUtf8Seq(ByteBufUtil.java:681)
[E] 	at io.netty.buffer.ByteBufUtil.writeUtf8(ByteBufUtil.java:638)
[E] 	at io.netty.buffer.Unpooled.copiedBufferUtf8(Unpooled.java:599)
[E] 	at io.netty.buffer.Unpooled.copiedBuffer(Unpooled.java:582)
[E] 	at zhttp.http.Response$HttpResponse.toJFullHttpResponse(Response.scala:37)
[E] 	at zhttp.service.server.ServerRequestHandler.writeAndFlush(ServerRequestHandler.scala:60)
[E] 	at zhttp.service.server.ServerRequestHandler.channelRead0(ServerRequestHandler.scala:76)
[E] 	at zhttp.service.server.ServerRequestHandler.channelRead0(ServerRequestHandler.scala:13)
[E] 	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
[E] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
[E] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
[E] 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
[E] 	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
[E] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
[E] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
[E] 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
[E] 	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
[E] 	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
[E] 	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
[E] 	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
[E] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
[E] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
[E] 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
[E] 	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
[E] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
[E] 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
[E] 	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
[E] 	at io.netty.channel.kqueue.AbstractKQueueStreamChannel$KQueueStreamUnsafe.readReady(AbstractKQueueStreamChannel.java:544)
[E] 	at io.netty.channel.kqueue.AbstractKQueueChannel$AbstractKQueueUnsafe.readReady(AbstractKQueueChannel.java:382)
[E] 	at io.netty.channel.kqueue.KQueueEventLoop.processReady(KQueueEventLoop.java:211)
[E] 	at io.netty.channel.kqueue.KQueueEventLoop.run(KQueueEventLoop.java:289)
[E] 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
[E] 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
[E] 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
[E] 	at java.base/java.lang.Thread.run(Thread.java:834)
[E]
```


</details>

For the whole time of the upload, `out.zip`'s size never changed (0B) and the application kept consuming memory (I can now see the JVM use 7GB according to macOS's activity monitor, 24GB according to `htop` and `bottom`). Reminder: the file is <1GB.

jconsole:

![image](https://user-images.githubusercontent.com/894884/111244542-6f952000-8603-11eb-916a-07289ba6efa6.png)

Note that this took around 37 seconds to send the payload.

For comparison, http4s 0.21.18:

```scala
import scala.concurrent.ExecutionContext

import cats.effect.ExitCode
import cats.effect.IO
import cats.effect.IOApp
import org.http4s._
import org.http4s.dsl.io._
import org.http4s.implicits._
import org.http4s.server.blaze.BlazeServerBuilder

object Main extends IOApp {
  def run(args: List[String]): IO[ExitCode] =
    BlazeServerBuilder[IO](ExecutionContext.global)
      .bindHttp(8090, "0.0.0.0")
      .withHttpApp(
        HttpRoutes
          .of[IO] { case req @ POST -> Root / "echo" =>
            Ok(req.body)
          }
          .orNotFound
      )
      .resource
      .use(_ => IO.never)
}
```

```
> curl localhost:8090/echo --data-binary @in.zip -v > out.zip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8090 (#0)
> POST /echo HTTP/1.1
> Host: localhost:8090
> User-Agent: curl/7.64.1
> Accept: */*
> Content-Length: 1000000000
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
>
< HTTP/1.1 200 OK
< Date: Tue, 16 Mar 2021 02:03:21 GMT
< Transfer-Encoding: chunked
<
* Done waiting for 100-continue
  0  953M    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0} [65536 bytes data]
 63 1401M    0  447M   46  448M   152M   152M  0:00:06  0:00:02  0:00:04  305M* We are completely uploaded and fine
{ [65552 bytes data]
100 1907M    0  953M  100  953M   248M   248M  0:00:03  0:00:03 --:--:--  497M
* Connection #0 to host localhost left intact
* Closing connection 0
*

```

(3 seconds)

```
> shasum --algorithm 256 *.zip
b021c1864684a4133e918330979f406fd372da0a2aeb94cb86d1cf9802d181aa  in.zip
b021c1864684a4133e918330979f406fd372da0a2aeb94cb86d1cf9802d181aa  out.zip
```

**Comment by kubukoz** [comment]
I can see

```scala
/**
 * Content holder for Requests and Responses
 */
sealed trait HttpContent[-R, +A] extends Product with Serializable

object HttpContent {
  case object Empty                                            extends HttpContent[Any, Nothing]
  final case class Complete[A](data: A)                        extends HttpContent[Any, A]
  final case class Chunked[R, A](data: ZStream[R, Nothing, A]) extends HttpContent[R, A]
}
```

but `Chunked` isn't used anywhere, so I guess streaming/chunked data is just not supported yet?

**Comment by amitksingh1490** [comment]
Yes, it's not supported yet. We are working on it

**Comment by tusharmath** [comment]
Thanks for such a detailed repro @kubukoz 🙏 

I am not sure if this issue is because of chunking since you are passing content-length in your request. I'll take a look at it in more detail and post my findings.

**Comment by tusharmath** [comment]
Temp Fix in #60
Will be throwing a 413 error in case the request is greater than 4bytes (configurable)
